# Comparing `tmp/pciSeq-0.0.57.dev0-py3-none-any.whl.zip` & `tmp/pciSeq-0.0.57.dev1-py3-none-any.whl.zip`

## zipinfo {}

```diff
@@ -1,75 +1,75 @@
-Zip file size: 366278 bytes, number of entries: 73
--rw-rw-rw-  2.0 fat     2135 b- defN 23-Jul-09 15:28 pciSeq/__init__.py
--rw-rw-rw-  2.0 fat    11165 b- defN 23-Jul-09 18:54 pciSeq/app.py
--rw-rw-rw-  2.0 fat     3713 b- defN 23-Jul-09 19:18 pciSeq/config.py
+Zip file size: 366076 bytes, number of entries: 73
+-rw-rw-rw-  2.0 fat     2135 b- defN 23-Jul-22 12:16 pciSeq/__init__.py
+-rw-rw-rw-  2.0 fat    10969 b- defN 23-Jul-23 11:30 pciSeq/app.py
+-rw-rw-rw-  2.0 fat     3713 b- defN 23-Jul-22 12:17 pciSeq/config.py
 -rw-rw-rw-  2.0 fat     3678 b- defN 22-Apr-07 23:15 pciSeq/make_plot.py
 -rw-rw-rw-  2.0 fat        0 b- defN 22-Jan-19 17:53 pciSeq/src/__init__.py
--rw-rw-rw-  2.0 fat       27 b- defN 23-Jul-09 19:27 pciSeq/src/_version.py
+-rw-rw-rw-  2.0 fat       27 b- defN 23-Jul-23 11:34 pciSeq/src/_version.py
 -rw-rw-rw-  2.0 fat        0 b- defN 22-Jan-19 17:53 pciSeq/src/cell_call/__init__.py
--rw-rw-rw-  2.0 fat    22062 b- defN 23-Jul-09 18:34 pciSeq/src/cell_call/datatypes.py
--rw-rw-rw-  2.0 fat     2995 b- defN 23-Jul-09 15:28 pciSeq/src/cell_call/log_config.py
--rw-rw-rw-  2.0 fat    23046 b- defN 23-Jul-09 18:41 pciSeq/src/cell_call/main.py
--rw-rw-rw-  2.0 fat     3104 b- defN 23-Jul-09 15:28 pciSeq/src/cell_call/summary.py
--rw-rw-rw-  2.0 fat     9194 b- defN 23-Jul-09 15:28 pciSeq/src/cell_call/utils.py
--rw-rw-rw-  2.0 fat        0 b- defN 23-Jul-09 15:28 pciSeq/src/diagnostics/__init__.py
--rw-rw-rw-  2.0 fat      299 b- defN 23-Jul-09 15:28 pciSeq/src/diagnostics/config.py
--rw-rw-rw-  2.0 fat     3336 b- defN 23-Jul-09 15:28 pciSeq/src/diagnostics/diagnostics.py
--rw-rw-rw-  2.0 fat      796 b- defN 23-Jul-09 15:28 pciSeq/src/diagnostics/launch_diagnostics.py
--rw-rw-rw-  2.0 fat     7340 b- defN 23-Jul-09 17:51 pciSeq/src/diagnostics/utils.py
+-rw-rw-rw-  2.0 fat    22062 b- defN 23-Jul-22 12:17 pciSeq/src/cell_call/datatypes.py
+-rw-rw-rw-  2.0 fat     2995 b- defN 23-Jul-22 12:13 pciSeq/src/cell_call/log_config.py
+-rw-rw-rw-  2.0 fat    22954 b- defN 23-Jul-22 12:17 pciSeq/src/cell_call/main.py
+-rw-rw-rw-  2.0 fat     3104 b- defN 23-Jul-22 12:13 pciSeq/src/cell_call/summary.py
+-rw-rw-rw-  2.0 fat     9194 b- defN 23-Jul-22 12:16 pciSeq/src/cell_call/utils.py
+-rw-rw-rw-  2.0 fat        0 b- defN 23-Jul-22 12:16 pciSeq/src/diagnostics/__init__.py
+-rw-rw-rw-  2.0 fat      299 b- defN 23-Jul-22 12:16 pciSeq/src/diagnostics/config.py
+-rw-rw-rw-  2.0 fat     3336 b- defN 23-Jul-22 12:16 pciSeq/src/diagnostics/diagnostics.py
+-rw-rw-rw-  2.0 fat      796 b- defN 23-Jul-22 12:16 pciSeq/src/diagnostics/launch_diagnostics.py
+-rw-rw-rw-  2.0 fat     7340 b- defN 23-Jul-22 12:17 pciSeq/src/diagnostics/utils.py
 -rw-rw-rw-  2.0 fat        0 b- defN 22-Jan-19 17:53 pciSeq/src/preprocess/__init__.py
--rw-rw-rw-  2.0 fat    11253 b- defN 23-Jul-09 15:28 pciSeq/src/preprocess/cell_borders.py
+-rw-rw-rw-  2.0 fat    12947 b- defN 23-Jul-23 11:18 pciSeq/src/preprocess/cell_borders.py
 -rw-rw-rw-  2.0 fat     5323 b- defN 23-Jun-25 19:58 pciSeq/src/preprocess/regionprops.py
 -rw-rw-rw-  2.0 fat     6973 b- defN 23-Apr-26 22:13 pciSeq/src/preprocess/segmentation.py
--rw-rw-rw-  2.0 fat     4585 b- defN 23-Jul-09 17:51 pciSeq/src/preprocess/spot_labels.py
--rw-rw-rw-  2.0 fat       88 b- defN 23-Jul-09 17:51 pciSeq/src/preprocess/utils.py
+-rw-rw-rw-  2.0 fat     4653 b- defN 23-Jul-23 11:18 pciSeq/src/preprocess/spot_labels.py
+-rw-rw-rw-  2.0 fat       88 b- defN 23-Jul-22 12:17 pciSeq/src/preprocess/utils.py
 -rw-rw-rw-  2.0 fat        0 b- defN 22-Feb-20 14:12 pciSeq/src/viewer/__init__.py
--rw-rw-rw-  2.0 fat     2256 b- defN 23-Jul-09 15:28 pciSeq/src/viewer/run_flask.py
--rw-rw-rw-  2.0 fat     5007 b- defN 23-Jul-09 15:28 pciSeq/src/viewer/stage_image.py
--rw-rw-rw-  2.0 fat     9109 b- defN 23-Jul-09 17:51 pciSeq/src/viewer/utils.py
--rw-rw-rw-  2.0 fat    10895 b- defN 23-Jul-09 15:28 pciSeq/static/2D/index.html
--rw-rw-rw-  2.0 fat     6506 b- defN 23-Jul-09 15:28 pciSeq/static/2D/viewer/genes_datatable.html
--rw-rw-rw-  2.0 fat     9622 b- defN 23-Jul-09 15:28 pciSeq/static/2D/viewer/css/index.css
--rw-rw-rw-  2.0 fat     3135 b- defN 23-Jul-09 15:28 pciSeq/static/2D/viewer/css/progress.css
--rw-rw-rw-  2.0 fat     7716 b- defN 23-Jul-09 17:51 pciSeq/static/2D/viewer/js/classConfig.js
--rw-rw-rw-  2.0 fat     2314 b- defN 23-Jul-09 17:51 pciSeq/static/2D/viewer/js/config.js
--rw-rw-rw-  2.0 fat     5589 b- defN 23-Jul-09 15:28 pciSeq/static/2D/viewer/js/customControl.js
--rw-rw-rw-  2.0 fat    23501 b- defN 23-Jul-09 17:51 pciSeq/static/2D/viewer/js/dapi.js
--rw-rw-rw-  2.0 fat     5984 b- defN 23-Jul-09 15:28 pciSeq/static/2D/viewer/js/dataLoader.js
--rw-rw-rw-  2.0 fat    17232 b- defN 23-Jul-09 15:28 pciSeq/static/2D/viewer/js/donut.js
--rw-rw-rw-  2.0 fat     4106 b- defN 23-Jul-09 15:28 pciSeq/static/2D/viewer/js/dt.js
--rw-rw-rw-  2.0 fat     9835 b- defN 23-Jul-09 15:28 pciSeq/static/2D/viewer/js/glyphConfig.js
--rw-rw-rw-  2.0 fat     4389 b- defN 23-Jul-09 15:28 pciSeq/static/2D/viewer/js/glyphPaths.js
--rw-rw-rw-  2.0 fat     7224 b- defN 23-Jul-09 15:28 pciSeq/static/2D/viewer/js/glyphs.js
--rw-rw-rw-  2.0 fat    12281 b- defN 23-Jul-09 15:28 pciSeq/static/2D/viewer/js/index.js
--rw-rw-rw-  2.0 fat     1073 b- defN 23-Jul-09 15:28 pciSeq/static/2D/viewer/js/progress.js
--rw-rw-rw-  2.0 fat     6367 b- defN 23-Jul-09 15:28 pciSeq/static/2D/viewer/js/stage_cells.js
--rw-rw-rw-  2.0 fat    12064 b- defN 23-Jul-09 15:28 pciSeq/static/2D/viewer/js/stage_glyphs.js
--rw-rw-rw-  2.0 fat     9540 b- defN 23-Jul-09 15:28 pciSeq/static/2D/viewer/js/stage_markers.js
--rw-rw-rw-  2.0 fat     7580 b- defN 23-Jul-09 17:51 pciSeq/static/2D/viewer/js/stage_markers_patched.js
--rw-rw-rw-  2.0 fat    15523 b- defN 23-Jul-09 15:28 pciSeq/static/2D/viewer/js/stage_polygons.js
--rw-rw-rw-  2.0 fat     4471 b- defN 23-Jul-09 15:28 pciSeq/static/2D/viewer/js/streaming-tsv-parser.js
--rw-rw-rw-  2.0 fat     4692 b- defN 23-Jul-09 15:28 pciSeq/static/2D/viewer/js/viewerUtils.js
--rw-rw-rw-  2.0 fat      995 b- defN 23-Jul-09 15:28 pciSeq/static/2D/viewer/js/lib/css/L.Control.Layers.Tree.css
--rw-rw-rw-  2.0 fat      421 b- defN 23-Jul-09 15:28 pciSeq/static/2D/viewer/js/lib/css/Leaflet.Coordinates-0.1.3.css
--rw-rw-rw-  2.0 fat   122544 b- defN 23-Jul-09 15:28 pciSeq/static/2D/viewer/js/lib/css/bootstrap.min.css
--rw-rw-rw-  2.0 fat     8669 b- defN 23-Jul-09 15:28 pciSeq/static/2D/viewer/js/lib/css/index.css
--rw-rw-rw-  2.0 fat     1228 b- defN 23-Jul-09 15:28 pciSeq/static/2D/viewer/js/lib/css/keen-dashboards.css
--rw-rw-rw-  2.0 fat     5245 b- defN 23-Jul-09 15:28 pciSeq/static/2D/viewer/js/lib/css/screen.css
--rw-rw-rw-  2.0 fat    26652 b- defN 23-Jul-09 15:28 pciSeq/static/2D/viewer/js/lib/js/L.Control.Layers.Tree.js
--rw-rw-rw-  2.0 fat     8813 b- defN 23-Jul-09 15:28 pciSeq/static/2D/viewer/js/lib/js/Leaflet.Coordinates-0.1.3.src.js
--rw-rw-rw-  2.0 fat    31824 b- defN 23-Jul-09 15:28 pciSeq/static/2D/viewer/js/lib/js/bootstrap.min.js
--rw-rw-rw-  2.0 fat    42707 b- defN 23-Jul-09 15:28 pciSeq/static/2D/viewer/js/lib/js/keen.min.js
--rw-rw-rw-  2.0 fat    15374 b- defN 23-Jul-09 15:28 pciSeq/static/2D/viewer/js/lib/js/leaflet-pip.js
--rw-rw-rw-  2.0 fat     5927 b- defN 23-Jul-09 15:28 pciSeq/static/2D/viewer/js/lib/js/leaflet.textpath.js
--rw-rw-rw-  2.0 fat      265 b- defN 23-Jul-09 15:28 pciSeq/static/2D/viewer/js/lib/js/preloader.js
--rw-rw-rw-  2.0 fat     9097 b- defN 23-Jul-09 15:28 pciSeq/static/2D/viewer/js/lib/js/pixiOverlay/MarkerContainer.js
--rw-rw-rw-  2.0 fat     4434 b- defN 23-Jul-09 15:28 pciSeq/static/2D/viewer/js/lib/js/pixiOverlay/bezier-easing.js
--rw-rw-rw-  2.0 fat   613808 b- defN 23-Jul-09 15:28 pciSeq/static/2D/viewer/js/lib/js/pixiOverlay/example.min.js
--rw-rw-rw-  2.0 fat    91874 b- defN 23-Jul-09 15:28 pciSeq/static/2D/viewer/js/lib/js/pixiOverlay/tools.min.js
--rw-rw-rw-  2.0 fat     1091 b- defN 23-Jul-09 19:36 pciSeq-0.0.57.dev0.dist-info/LICENCE
--rw-rw-rw-  2.0 fat     5279 b- defN 23-Jul-09 19:36 pciSeq-0.0.57.dev0.dist-info/METADATA
--rw-rw-rw-  2.0 fat       92 b- defN 23-Jul-09 19:36 pciSeq-0.0.57.dev0.dist-info/WHEEL
--rw-rw-rw-  2.0 fat       48 b- defN 23-Jul-09 19:36 pciSeq-0.0.57.dev0.dist-info/entry_points.txt
--rw-rw-rw-  2.0 fat        7 b- defN 23-Jul-09 19:36 pciSeq-0.0.57.dev0.dist-info/top_level.txt
-?rw-rw-r--  2.0 fat     6965 b- defN 23-Jul-09 19:36 pciSeq-0.0.57.dev0.dist-info/RECORD
-73 files, 1332482 bytes uncompressed, 354968 bytes compressed:  73.4%
+-rw-rw-rw-  2.0 fat     2191 b- defN 23-Jul-22 12:17 pciSeq/src/viewer/run_flask.py
+-rw-rw-rw-  2.0 fat     4047 b- defN 23-Jul-22 12:17 pciSeq/src/viewer/stage_image.py
+-rw-rw-rw-  2.0 fat     9109 b- defN 23-Jul-22 12:17 pciSeq/src/viewer/utils.py
+-rw-rw-rw-  2.0 fat    10895 b- defN 23-Jul-22 12:13 pciSeq/static/2D/index.html
+-rw-rw-rw-  2.0 fat     6506 b- defN 23-Jul-22 12:13 pciSeq/static/2D/viewer/genes_datatable.html
+-rw-rw-rw-  2.0 fat     9622 b- defN 23-Jul-22 12:13 pciSeq/static/2D/viewer/css/index.css
+-rw-rw-rw-  2.0 fat     3135 b- defN 23-Jul-22 12:13 pciSeq/static/2D/viewer/css/progress.css
+-rw-rw-rw-  2.0 fat     7716 b- defN 23-Jul-22 12:17 pciSeq/static/2D/viewer/js/classConfig.js
+-rw-rw-rw-  2.0 fat     2314 b- defN 23-Jul-22 12:17 pciSeq/static/2D/viewer/js/config.js
+-rw-rw-rw-  2.0 fat     5589 b- defN 23-Jul-22 12:13 pciSeq/static/2D/viewer/js/customControl.js
+-rw-rw-rw-  2.0 fat    23501 b- defN 23-Jul-22 12:17 pciSeq/static/2D/viewer/js/dapi.js
+-rw-rw-rw-  2.0 fat     5984 b- defN 23-Jul-22 12:13 pciSeq/static/2D/viewer/js/dataLoader.js
+-rw-rw-rw-  2.0 fat    17232 b- defN 23-Jul-22 12:13 pciSeq/static/2D/viewer/js/donut.js
+-rw-rw-rw-  2.0 fat     4106 b- defN 23-Jul-22 12:13 pciSeq/static/2D/viewer/js/dt.js
+-rw-rw-rw-  2.0 fat     9835 b- defN 23-Jul-22 12:13 pciSeq/static/2D/viewer/js/glyphConfig.js
+-rw-rw-rw-  2.0 fat     4389 b- defN 23-Jul-22 12:13 pciSeq/static/2D/viewer/js/glyphPaths.js
+-rw-rw-rw-  2.0 fat     7224 b- defN 23-Jul-22 12:13 pciSeq/static/2D/viewer/js/glyphs.js
+-rw-rw-rw-  2.0 fat    12281 b- defN 23-Jul-22 12:13 pciSeq/static/2D/viewer/js/index.js
+-rw-rw-rw-  2.0 fat     1073 b- defN 23-Jul-22 12:13 pciSeq/static/2D/viewer/js/progress.js
+-rw-rw-rw-  2.0 fat     6367 b- defN 23-Jul-22 12:13 pciSeq/static/2D/viewer/js/stage_cells.js
+-rw-rw-rw-  2.0 fat    12064 b- defN 23-Jul-22 12:13 pciSeq/static/2D/viewer/js/stage_glyphs.js
+-rw-rw-rw-  2.0 fat     9540 b- defN 23-Jul-22 12:13 pciSeq/static/2D/viewer/js/stage_markers.js
+-rw-rw-rw-  2.0 fat     7580 b- defN 23-Jul-22 12:17 pciSeq/static/2D/viewer/js/stage_markers_patched.js
+-rw-rw-rw-  2.0 fat    15523 b- defN 23-Jul-22 12:13 pciSeq/static/2D/viewer/js/stage_polygons.js
+-rw-rw-rw-  2.0 fat     4471 b- defN 23-Jul-22 12:13 pciSeq/static/2D/viewer/js/streaming-tsv-parser.js
+-rw-rw-rw-  2.0 fat     4692 b- defN 23-Jul-22 12:13 pciSeq/static/2D/viewer/js/viewerUtils.js
+-rw-rw-rw-  2.0 fat      995 b- defN 23-Jul-22 12:13 pciSeq/static/2D/viewer/js/lib/css/L.Control.Layers.Tree.css
+-rw-rw-rw-  2.0 fat      421 b- defN 23-Jul-22 12:13 pciSeq/static/2D/viewer/js/lib/css/Leaflet.Coordinates-0.1.3.css
+-rw-rw-rw-  2.0 fat   122544 b- defN 23-Jul-22 12:13 pciSeq/static/2D/viewer/js/lib/css/bootstrap.min.css
+-rw-rw-rw-  2.0 fat     8669 b- defN 23-Jul-22 12:13 pciSeq/static/2D/viewer/js/lib/css/index.css
+-rw-rw-rw-  2.0 fat     1228 b- defN 23-Jul-22 12:13 pciSeq/static/2D/viewer/js/lib/css/keen-dashboards.css
+-rw-rw-rw-  2.0 fat     5245 b- defN 23-Jul-22 12:13 pciSeq/static/2D/viewer/js/lib/css/screen.css
+-rw-rw-rw-  2.0 fat    26652 b- defN 23-Jul-22 12:13 pciSeq/static/2D/viewer/js/lib/js/L.Control.Layers.Tree.js
+-rw-rw-rw-  2.0 fat     8813 b- defN 23-Jul-22 12:13 pciSeq/static/2D/viewer/js/lib/js/Leaflet.Coordinates-0.1.3.src.js
+-rw-rw-rw-  2.0 fat    31824 b- defN 23-Jul-22 12:13 pciSeq/static/2D/viewer/js/lib/js/bootstrap.min.js
+-rw-rw-rw-  2.0 fat    42707 b- defN 23-Jul-22 12:13 pciSeq/static/2D/viewer/js/lib/js/keen.min.js
+-rw-rw-rw-  2.0 fat    15374 b- defN 23-Jul-22 12:13 pciSeq/static/2D/viewer/js/lib/js/leaflet-pip.js
+-rw-rw-rw-  2.0 fat     5927 b- defN 23-Jul-22 12:13 pciSeq/static/2D/viewer/js/lib/js/leaflet.textpath.js
+-rw-rw-rw-  2.0 fat      265 b- defN 23-Jul-22 12:13 pciSeq/static/2D/viewer/js/lib/js/preloader.js
+-rw-rw-rw-  2.0 fat     9097 b- defN 23-Jul-22 12:13 pciSeq/static/2D/viewer/js/lib/js/pixiOverlay/MarkerContainer.js
+-rw-rw-rw-  2.0 fat     4434 b- defN 23-Jul-22 12:13 pciSeq/static/2D/viewer/js/lib/js/pixiOverlay/bezier-easing.js
+-rw-rw-rw-  2.0 fat   613808 b- defN 23-Jul-22 12:13 pciSeq/static/2D/viewer/js/lib/js/pixiOverlay/example.min.js
+-rw-rw-rw-  2.0 fat    91874 b- defN 23-Jul-22 12:13 pciSeq/static/2D/viewer/js/lib/js/pixiOverlay/tools.min.js
+-rw-rw-rw-  2.0 fat     1091 b- defN 23-Jul-23 11:37 pciSeq-0.0.57.dev1.dist-info/LICENCE
+-rw-rw-rw-  2.0 fat     5279 b- defN 23-Jul-23 11:37 pciSeq-0.0.57.dev1.dist-info/METADATA
+-rw-rw-rw-  2.0 fat       92 b- defN 23-Jul-23 11:37 pciSeq-0.0.57.dev1.dist-info/WHEEL
+-rw-rw-rw-  2.0 fat       48 b- defN 23-Jul-23 11:37 pciSeq-0.0.57.dev1.dist-info/entry_points.txt
+-rw-rw-rw-  2.0 fat        7 b- defN 23-Jul-23 11:37 pciSeq-0.0.57.dev1.dist-info/top_level.txt
+?rw-rw-r--  2.0 fat     6965 b- defN 23-Jul-23 11:37 pciSeq-0.0.57.dev1.dist-info/RECORD
+73 files, 1332931 bytes uncompressed, 354766 bytes compressed:  73.4%
```

## zipnote {}

```diff
@@ -195,26 +195,26 @@
 
 Filename: pciSeq/static/2D/viewer/js/lib/js/pixiOverlay/example.min.js
 Comment: 
 
 Filename: pciSeq/static/2D/viewer/js/lib/js/pixiOverlay/tools.min.js
 Comment: 
 
-Filename: pciSeq-0.0.57.dev0.dist-info/LICENCE
+Filename: pciSeq-0.0.57.dev1.dist-info/LICENCE
 Comment: 
 
-Filename: pciSeq-0.0.57.dev0.dist-info/METADATA
+Filename: pciSeq-0.0.57.dev1.dist-info/METADATA
 Comment: 
 
-Filename: pciSeq-0.0.57.dev0.dist-info/WHEEL
+Filename: pciSeq-0.0.57.dev1.dist-info/WHEEL
 Comment: 
 
-Filename: pciSeq-0.0.57.dev0.dist-info/entry_points.txt
+Filename: pciSeq-0.0.57.dev1.dist-info/entry_points.txt
 Comment: 
 
-Filename: pciSeq-0.0.57.dev0.dist-info/top_level.txt
+Filename: pciSeq-0.0.57.dev1.dist-info/top_level.txt
 Comment: 
 
-Filename: pciSeq-0.0.57.dev0.dist-info/RECORD
+Filename: pciSeq-0.0.57.dev1.dist-info/RECORD
 Comment: 
 
 Zip file comment:
```

## pciSeq/app.py

```diff
@@ -262,25 +262,21 @@
 
 if __name__ == "__main__":
     # set up the logger
     attach_to_log()
 
     # read some demo data
     _iss_spots = pd.read_csv(os.path.join(ROOT_DIR, 'data', 'mouse', 'ca1', 'iss', 'spots.csv'))
-    spots = pd.read_csv(r"F:\data\Sara\20230707\spotsT007.csv")
 
-    dat = np.load(r'F:\data\Sara\20230707\background_image_fixed_seg 2.npy', allow_pickle=True).item()
-    masks = dat['masks']
-    coo = coo_matrix(masks)
-    # _coo = load_npz(os.path.join(ROOT_DIR, 'data', 'mouse', 'ca1', 'segmentation', 'label_image.coo.npz'))
-    #
-    # _scRNAseq = pd.read_csv(os.path.join(ROOT_DIR, 'data', 'mouse', 'ca1', 'scRNA', 'scRNAseq.csv.gz'),
-    #                         header=None, index_col=0, compression='gzip', dtype=object)
-    # _scRNAseq = _scRNAseq.rename(columns=_scRNAseq.iloc[0], copy=False).iloc[1:]
-    # _scRNAseq = _scRNAseq.astype(float).astype(np.uint32)
+    _coo = load_npz(os.path.join(ROOT_DIR, 'data', 'mouse', 'ca1', 'segmentation', 'label_image.coo.npz'))
+
+    _scRNAseq = pd.read_csv(os.path.join(ROOT_DIR, 'data', 'mouse', 'ca1', 'scRNA', 'scRNAseq.csv.gz'),
+                            header=None, index_col=0, compression='gzip', dtype=object)
+    _scRNAseq = _scRNAseq.rename(columns=_scRNAseq.iloc[0], copy=False).iloc[1:]
+    _scRNAseq = _scRNAseq.astype(float).astype(np.uint32)
 
     # main task
     # _opts = {'max_iter': 10}
-    fit(spots=spots, coo=coo, opts={'save_data': True,'launch_viewer': True, 'launch_diagnostics': True})
-
+    _opts = {'save_data': True,'launch_viewer': True, 'launch_diagnostics': True}
+    fit(spots=_iss_spots, coo=_coo, scRNAseq=_scRNAseq, opts=_opts)
```

## pciSeq/src/_version.py

```diff
@@ -1 +1 @@
-__version__ = '0.0.57.dev0'
+__version__ = '0.0.57.dev1'
```

## pciSeq/src/cell_call/main.py

```diff
@@ -62,31 +62,27 @@
 
             # 1. For each cell, calc the expected gene counts
             self.geneCount_upd()
 
             # 2. calc expected gamma
             self.gamma_upd()
 
-            if self.single_cell.isMissing:
-                self.gaussian_upd()
-
             # 3. assign cells to cell types
             self.cell_to_cellType()
 
-            if self.single_cell.isMissing:
-                self.dalpha_upd()
-
             # 4. assign spots to cells
             self.spots_to_cell()
 
             # 5. update gene efficiency
             self.eta_upd()
 
             # 6. Update single cell data
             if self.single_cell.isMissing:
+                self.gaussian_upd()
+                self.dalpha_upd()
                 self.mu_upd()
 
             self.has_converged, delta = utils.hasConverged(self.spots, p0, self.config['CellCallTolerance'])
             logger.info(' Iteration %d, mean prob change %f' % (i, delta))
 
             # replace p0 with the latest probabilities
             p0 = self.spots.parent_cell_prob
```

## pciSeq/src/preprocess/cell_borders.py

```diff
@@ -1,11 +1,13 @@
 """ Functions to extract the cell boundaries """
 import pandas as pd
 import numpy as np
 import diplib as dip
+from multiprocessing import Pool, cpu_count
+from multiprocessing.dummy import Pool as ThreadPool
 
 
 def extract_borders_dip(label_image, offset_x=0, offset_y=0, exclude_labels=(0,)):
     """
     Extracts the cell boundaries from the label image array. The background is
     assumed to have label=0 and it will be ignored by default.
     Parameters
@@ -40,14 +42,61 @@
             pass
     df = pd.DataFrame([d]).T
     df = df.reset_index()
     df.columns = ['label', 'coords']
     return df
 
 
+def extract_borders(cell_labels):
+    '''
+    Extracts the cell boundaries from the label image array. Same as 'extract_borders_dip()' but a lot faster.
+    Parameters
+    ----------
+    label_image:    The label image array, typically obtained from some image segmentation
+                    application and maps every pixel on the image to a cell label.
+
+    Returns
+    -------
+    Returns a dataframe with columns 'labels' and 'coords'
+    """
+    '''
+    cell_boundaries = pd.DataFrame()
+    borders_list = _extract_borders(cell_labels)
+    d = dict(borders_list)
+    cell_boundaries['label'] = d.keys()
+    cell_boundaries['coords'] = d.values()
+    return cell_boundaries
+
+
+def _extract_borders(label_image):
+    """
+    Extracts the cell boundaries from the label image array.
+    Returns a dict where keys are the cell label and values the corresponding cell boundaries
+    """
+
+    # labels = sorted(set(label_image.flatten()) - set(exclude_labels))
+    cc = dip.GetImageChainCodes(label_image)  # input must be an unsigned integer type
+
+    pool = ThreadPool(cpu_count())
+    # it would be nice to process only those cc whose cc.objectID is in labels
+    results = pool.map(parse_chaincode, cc)
+    # close the pool and wait for the work to finish
+    pool.close()
+    pool.join()
+
+    return dict(results)
+
+
+def parse_chaincode(c):
+    p = c.Polygon().Simplify()
+    p = p + np.array([0, 0])
+    p = np.uint64(p).tolist()
+    p.append(p[0])  # append the first pair at the end to close the polygon
+    return np.uint64(c.objectID), p
+
 # *********************************************************************************
 # All Functions below are all deprecated. WIll be cleaned-p at a later verion
 # *********************************************************************************
 #
 # def cell_boundaries(stage, cell_props):
 #     '''
 #     calculate the outlines of the cells
```

## pciSeq/src/preprocess/spot_labels.py

```diff
@@ -5,15 +5,15 @@
 """
 
 import numpy as np
 import pandas as pd
 import skimage.measure as skmeas
 from typing import Tuple
 from scipy.sparse import coo_matrix, csr_matrix
-from pciSeq.src.preprocess.cell_borders import extract_borders_dip
+from pciSeq.src.preprocess.cell_borders import extract_borders_dip, extract_borders
 from pciSeq.src.cell_call.log_config import logger
 
 
 def inside_cell(label_image, spots) -> np.array:
     if isinstance(label_image, coo_matrix):
         label_image = label_image.tocsr()
     elif isinstance(label_image, np.ndarray):
@@ -74,26 +74,25 @@
     spots = spots[mask_x & mask_y]
 
     # 1. Find which cell the spots lie within
     inc = inside_cell(coo.tocsr(), spots)
     spots = spots.assign(label=inc)
 
     # 2. Get cell centroids and area
-    props = skmeas.regionprops(coo.toarray().astype(np.int32))
-    props_df = pd.DataFrame(data=[(d.label, d.area, d.centroid[1], d.centroid[0]) for d in props],
-                      columns=['label', 'area', 'x_cell', 'y_cell'])
+    props = skmeas.regionprops_table(coo.toarray().astype(np.int32), properties=['label', 'area', 'centroid'])
+    props_df = pd.DataFrame(props).rename(columns={'centroid-0': 'y_cell', 'centroid-1': 'x_cell'})
 
     # if there is a label map, attach it to the cell props.
     if label_map is not None:
         props_df = pd.merge(props_df, label_map, left_on='label', right_on='new_label', how='left')
         props_df = props_df.drop(['new_label'], axis=1)
 
     # 3. Get the cell boundaries
-    cell_boundaries = extract_borders_dip(coo.toarray().astype(np.uint32))
-
+    # cell_boundaries = extract_borders_dip(coo.toarray().astype(np.uint32))
+    cell_boundaries = extract_borders(coo.toarray().astype(np.uint32))
     assert props_df.shape[0] == cell_boundaries.shape[0] == np.unique(coo.data).shape[0]
     assert set(spots.label[spots.label > 0]) <= set(props_df.label)
 
     cells = props_df.merge(cell_boundaries)
     cells.sort_values(by=['label', 'x_cell', 'y_cell'])
     assert cells.shape[0] == cell_boundaries.shape[0] == props_df.shape[0]
```

## pciSeq/src/viewer/run_flask.py

```diff
@@ -1,11 +1,7 @@
-"""
-Function that will open an html page in the browser
-"""
-
 from flask import Flask, render_template
 import os
 import webbrowser
 import platform
 import random
 from threading import Timer
 from pciSeq.src.cell_call.log_config import logger
```

## pciSeq/src/viewer/stage_image.py

```diff
@@ -1,22 +1,13 @@
-''''
-splits the big background image  into smaller images size 2000x2000px.
-Look also at https://stackoverflow.com/questions/10853119/chop-image-into-tiles-using-vips-command-line/15293104
-for an alternative (and probably better way)
-It also creates the pyramid tiles for the viewer
-'''
 import shutil
 import os
 import pyvips
 from pciSeq.src.cell_call.log_config import logger
 
 
-# logger = logging.getLogger(__name__)
-
-
 def split_image(im):
     # DEPRECATED to be removed
     '''
     you can just do:
         im.dzsave('./out', suffix='.tif', skip_blanks=-1, background=0, depth='one', overlap=0, tile_size=2000, layout='google')
     to split the image to smaller squares. However you need to write a couple of line to rename and move the file to the correct
     folders
@@ -111,24 +102,7 @@
     im.dzsave(out_dir, layout='google', suffix='.jpg', background=0)
     logger.info('Done. Pyramid of tiles saved at: %s' % out_dir)
 
     return pixel_dims
 
 
 
-if __name__ == "__main__":
-    # imPath = r"F:\data\Sara\pciSeq\data\debug\lro_nearside.jpg"
-    imPath = r"F:\data\Sara\pciSeq\data\debug\1024px-Wac_nearside.jpg"
-    # imPath = r"F:\data\Sara\pciSeq\data\debug\tdTomato_image_fixed.jpg"
-    # imPath = r"/mnt/f/data/Sara/pciSeq/data/debug/tdTomato_image_fixed.jpg"
-    # split_image(im)
-
-    # # to rotate the image do:
-    # im = pyvips.Image.new_from_file(imPath)
-    # im = im.rotate(90, interpolate=pyvips.Interpolate.new("nearest"))
-    # im.write_to_file(r'data/background_image/background_image_adj_rot.tif')
-
-    tile_maker(imPath, z_depth=4)
-
-
-
-
```

## Comparing `pciSeq-0.0.57.dev0.dist-info/LICENCE` & `pciSeq-0.0.57.dev1.dist-info/LICENCE`

 * *Files identical despite different names*

## Comparing `pciSeq-0.0.57.dev0.dist-info/METADATA` & `pciSeq-0.0.57.dev1.dist-info/METADATA`

 * *Files 0% similar despite different names*

```diff
@@ -1,10 +1,10 @@
 Metadata-Version: 2.1
 Name: pciSeq
-Version: 0.0.57.dev0
+Version: 0.0.57.dev1
 Summary: Probabilistic cell typing for spatial transcriptomics
 Home-page: https://github.com/acycliq/pciSeq
 Author: Dimitris Nicoloutsopoulos
 Author-email: dimitris.nicoloutsopoulos@gmail.com
 License: BSD
 Classifier: Programming Language :: Python :: 3
 Classifier: License :: OSI Approved :: BSD License
```

## Comparing `pciSeq-0.0.57.dev0.dist-info/RECORD` & `pciSeq-0.0.57.dev1.dist-info/RECORD`

 * *Files 6% similar despite different names*

```diff
@@ -1,33 +1,33 @@
 pciSeq/__init__.py,sha256=mxqjE8smLqDQuwzheJ_iA3gfcyV7c4x8lHQPwSEQwJk,2135
-pciSeq/app.py,sha256=CaJQSw3_D8HOxy1yIFBvTYaW9hhObAtW0mDxsoqyIdY,11165
+pciSeq/app.py,sha256=PY6hbjBu6fvAcrRem5uzO3h3dUZ7fdAgjkd3TKK3w2c,10969
 pciSeq/config.py,sha256=UgT_7SEMMImfuMTIpCw3_t-DYMRMtbZ-SqBjq0rsGMQ,3713
 pciSeq/make_plot.py,sha256=2AkncrVVvytr2keK505oMC0LV1Txli9WpsSGpjmht_4,3678
 pciSeq/src/__init__.py,sha256=47DEQpj8HBSa-_TImW-5JCeuQeRkm5NMpJWZG3hSuFU,0
-pciSeq/src/_version.py,sha256=D0ZlQn6xfQdpbfv9NDewCTqsljokvldKNO7o1C7xkk4,27
+pciSeq/src/_version.py,sha256=hMAIWMvqhydza_QPdhiNE7RH6I1K62XW4F0JATg9Yqk,27
 pciSeq/src/cell_call/__init__.py,sha256=47DEQpj8HBSa-_TImW-5JCeuQeRkm5NMpJWZG3hSuFU,0
 pciSeq/src/cell_call/datatypes.py,sha256=tpTSWdEVL_JozrX9kx23AniU5TcMb-GFRBRg2HnOHNw,22062
 pciSeq/src/cell_call/log_config.py,sha256=vnm9TxlMEVG_GLPfl21GDEWTafEk3B-drTLsVQMfFEE,2995
-pciSeq/src/cell_call/main.py,sha256=0IA8rtIkeEiQ1nGfuN5vcxlayx1HcdAqdI_t5KodK7c,23046
+pciSeq/src/cell_call/main.py,sha256=hIoZTMF6R5HFo8IxoHNoNOa3ip4P6PcPdFCFycZboRI,22954
 pciSeq/src/cell_call/summary.py,sha256=f2snLqJIeywtYuHrJLM-hRvifgqRU2u2aCKE_JrZJPs,3104
 pciSeq/src/cell_call/utils.py,sha256=M4KJrS82pIHuCoxg2ej6NE9fJAVCvJw90T7YmWI2T7o,9194
 pciSeq/src/diagnostics/__init__.py,sha256=47DEQpj8HBSa-_TImW-5JCeuQeRkm5NMpJWZG3hSuFU,0
 pciSeq/src/diagnostics/config.py,sha256=kfPzvlFdGN0QgdfECwVV8lujXUiyim4fuetJoU4ziBQ,299
 pciSeq/src/diagnostics/diagnostics.py,sha256=L8tBgykN-KktrUS8YdIP5doi8meVVUUZU2PvOcUt72g,3336
 pciSeq/src/diagnostics/launch_diagnostics.py,sha256=p2t1wDc48iWvpfEjUt_-J5iK37EXPMg2O12y-pCeE_w,796
 pciSeq/src/diagnostics/utils.py,sha256=fNolyKf8SDl-W113NRxvEEwm9J5rxK_K_-a35YigI68,7340
 pciSeq/src/preprocess/__init__.py,sha256=47DEQpj8HBSa-_TImW-5JCeuQeRkm5NMpJWZG3hSuFU,0
-pciSeq/src/preprocess/cell_borders.py,sha256=eTuFuexqGMFEV4upuTRInoaAi75NN46TIcvJ36TbZ4Q,11253
+pciSeq/src/preprocess/cell_borders.py,sha256=FstDLwvlsk4-vxKhLeHCQiQNU6AhON6X0ei_PiZpINE,12947
 pciSeq/src/preprocess/regionprops.py,sha256=gpScubR5adkKL5g9uSir8ud_pehiblMnDGc592c2S7c,5323
 pciSeq/src/preprocess/segmentation.py,sha256=9fc47EsxeFQnNtbJ0vm-RqATsBBMP73J5chdRFJocd4,6973
-pciSeq/src/preprocess/spot_labels.py,sha256=ukYTnp0xnu9J4-bKCTLEDhHAfaQNCvE7teEFyU0OLRk,4585
+pciSeq/src/preprocess/spot_labels.py,sha256=xtA2MXtVPDVDkORv1nHClxyz9SzJeoOQ_NRkixV-SMQ,4653
 pciSeq/src/preprocess/utils.py,sha256=7UB-fOKsSY_a4MVPs7E6BoNZcwITqm2N2JFMZqNYbj0,88
 pciSeq/src/viewer/__init__.py,sha256=47DEQpj8HBSa-_TImW-5JCeuQeRkm5NMpJWZG3hSuFU,0
-pciSeq/src/viewer/run_flask.py,sha256=9HS_RW-O9xddAeiApd4m_w2JJmbqUGZ5lhQ-z_HUeVA,2256
-pciSeq/src/viewer/stage_image.py,sha256=0zvdVk_R52T43ueSt3K3DgdMZPUyJNxDSyG0Bni7O78,5007
+pciSeq/src/viewer/run_flask.py,sha256=wj_6Vv3YCeLXmt7tr_7yqSCfGoPHrGyyuu8Q3rfRqdI,2191
+pciSeq/src/viewer/stage_image.py,sha256=dtR4OF1Tw6RnLwPzCBBuNL8cyUK6AQi0olr5BMdnRAY,4047
 pciSeq/src/viewer/utils.py,sha256=VwHLGIVQJnaRYeDAjoIoP87O8qXkeS7H4IHJynQPjsw,9109
 pciSeq/static/2D/index.html,sha256=9aMR5hOaXugJpSOt3ypUY24z6blsfOS83Cd9LHbAHa0,10895
 pciSeq/static/2D/viewer/genes_datatable.html,sha256=v10lTjkXYeAv0DVuaWHOB6XgG887k-_ABXAgobDV-tY,6506
 pciSeq/static/2D/viewer/css/index.css,sha256=IntuFc7kNAaNkLSL3m_9fpnH67i5KJpNWJ0OIW4TmD4,9622
 pciSeq/static/2D/viewer/css/progress.css,sha256=0MfjX9kIwD4wBeC39LBi9bCvu4ZkHwg_vtbuUiy4CuI,3135
 pciSeq/static/2D/viewer/js/classConfig.js,sha256=QuGSikifUK9McS_rBVCqwsoCiUl-89mncmlyYauYMMo,7716
 pciSeq/static/2D/viewer/js/config.js,sha256=mhuRBtP_lLeEQzSWYb9D28C4dfufLu3KLFu2Edr0Zik,2314
@@ -61,13 +61,13 @@
 pciSeq/static/2D/viewer/js/lib/js/leaflet-pip.js,sha256=QqsilU0SHTaKA_qNSv_jEDH53cWO6T_RdforfybnbKM,15374
 pciSeq/static/2D/viewer/js/lib/js/leaflet.textpath.js,sha256=i5cmMPqpqXShZz0TvXU3ZQUi2Y07SIbIpMYbhPE8gWE,5927
 pciSeq/static/2D/viewer/js/lib/js/preloader.js,sha256=eaFWRf9yT-Td4C2t01P8fDEZYszYTaGDgQ1vjevNIDI,265
 pciSeq/static/2D/viewer/js/lib/js/pixiOverlay/MarkerContainer.js,sha256=Z8TADCqYEQr25OEF7wljsOVAOpx87UrjFPI2nvIty70,9097
 pciSeq/static/2D/viewer/js/lib/js/pixiOverlay/bezier-easing.js,sha256=yMoRDqJiglg0M0eWM25WGA5zzLWuJpOyC3a8aGFx0jw,4434
 pciSeq/static/2D/viewer/js/lib/js/pixiOverlay/example.min.js,sha256=bZEnK8XSnShrpLYtOS7Xq7wnp-nW_nMZh4aaDc1Gt6U,613808
 pciSeq/static/2D/viewer/js/lib/js/pixiOverlay/tools.min.js,sha256=A027Ub-oWXvKbz2B3pgMFS5xApmnEHXz9GQh79YZZnU,91874
-pciSeq-0.0.57.dev0.dist-info/LICENCE,sha256=6kbiFSfobTZ7beWiKnHpN902HgBx-Jzgcme0SvKqhKY,1091
-pciSeq-0.0.57.dev0.dist-info/METADATA,sha256=VxMViJO-Wjq9ryYkbdR9qSWasZOuHAtp2DuVeFyAhZM,5279
-pciSeq-0.0.57.dev0.dist-info/WHEEL,sha256=G16H4A3IeoQmnOrYV4ueZGKSjhipXx8zc8nu9FGlvMA,92
-pciSeq-0.0.57.dev0.dist-info/entry_points.txt,sha256=RGOkAd0S7rrTuEInj-jKu5B7630i4R3mK9v9vx6nqc8,48
-pciSeq-0.0.57.dev0.dist-info/top_level.txt,sha256=mSonMN0SaB51m5yLGm5hFE1Fthf-2U4sRKxSezY-JMk,7
-pciSeq-0.0.57.dev0.dist-info/RECORD,,
+pciSeq-0.0.57.dev1.dist-info/LICENCE,sha256=6kbiFSfobTZ7beWiKnHpN902HgBx-Jzgcme0SvKqhKY,1091
+pciSeq-0.0.57.dev1.dist-info/METADATA,sha256=4jlWgIgWXxWvXdabRou77c0UgDqetYWUcpiW0kK25eY,5279
+pciSeq-0.0.57.dev1.dist-info/WHEEL,sha256=G16H4A3IeoQmnOrYV4ueZGKSjhipXx8zc8nu9FGlvMA,92
+pciSeq-0.0.57.dev1.dist-info/entry_points.txt,sha256=RGOkAd0S7rrTuEInj-jKu5B7630i4R3mK9v9vx6nqc8,48
+pciSeq-0.0.57.dev1.dist-info/top_level.txt,sha256=mSonMN0SaB51m5yLGm5hFE1Fthf-2U4sRKxSezY-JMk,7
+pciSeq-0.0.57.dev1.dist-info/RECORD,,
```

